pinc/MessageCodes.inc

```diff
@@ -0,0 +1,44 @@
+<?php
+
+class MessageCodes
+{
+    public const PROJECT_NOT_IN_ROUND = "Project not in a round.";
+    public const PROJECT_NOT_AVAILABLE = 'Project not available.';
+    public const REQUESTOR_NOT_LOGGED_IN = "User not logged in.";
+    public const USER_NOT_QUALIFIED_FOR_ROUND = 'User not cleared for round.';
+    public const PROJECT_REQUIRES_MAINTENANCE = 'Project requires maintenance.';
+    public const PROJECT_NOT_IN_REQUIRED_ROUND = "Project not in required round.";
+    public const INVALID_ROUND = "Invalid round.";
+    public const INVALID_PROJECT_ID = "No such project ID";
+    public const NO_MORE_PAGES = "No more pages.";
+    public const NO_SUCH_PAGE = "No such page in project";
+    public const WRONG_PAGE_STATE = "Page not in an allowed state";
+    public const WRONG_PAGE_OWNER = "You do not own this page";
+    public const DAILY_LIMIT_EXCEEDED = "Daily page limit exceeded";
+    public const REACHED_BEGINNERS_QUOTA = "Reached 'Beginners Only' quota.";
+    public const REACHED_PROJECT_BEGINNERS_QUOTA = "Reached project 'Beginners Only' quota.";
+    public const NO_ACCESS_TO_BEGINERS_ONLY = "No access to 'beginner' projects.";
+    public const RESERVED_FOR_NEW_PROOFERS = "Reserved for new proofreaders.";
+    public const NO_ACCESS_TO_MENTORS_ONLY = "No access 'mentors only' projects.";
+}
+
+// localized messages only needed for cases which can occur in normal operation
+// other cases can only happen if the client app. is faulty
+class LocalizedException extends Exception
```

Comment:
The challenge with throwing a generalized Exception (which is effectively what LocalizedException is) is that a caller can't catch a specific exception and do something with it without catching all of the generalized exceptions and then checking the message string.

While it might seem verbose to code, making each message its own exception type extending a base exception and including its custom string would be better programatically.

These seem like they logically fall into exception with either:
* project
* page
* user access

The first ones could be exceptions that extend the existing `ProjectException` base exception (indeed we already have `InvalidProjectIDException` for an invalid project ID). The second could extend from a new `ProjectPageException` base exception.  And the third extend from a new `UserAccessException` base exception (probably have these defined in `User.inc`).

Reply:
I'll try that.